drivers/spi/spi_mcux_flexcomm.c

```diff
@@ -218,6 +245,372 @@ static int spi_mcux_configure(const struct device *dev,
+	} else {
+		blk_cfg->dest_address = (uint32_t)buf;
+	}
+
+	blk_cfg->source_address = (uint32_t)&base->FIFORD;
+
+	/* direction is given by the DT */
+	stream->dma_cfg.head_block = blk_cfg;
+	stream->dma_cfg.user_data = data;
+
+	/* Enables the DMA request from SPI rxFIFO */
+	base->FIFOCFG |= SPI_FIFOCFG_DMARX_MASK;
+
+	/* pass our client origin to the dma: data->dma_rx.channel */
+	ret = dma_config(data->dma_rx.dma_dev, data->dma_rx.channel,
+			&stream->dma_cfg);
+	/* the channel is the actual stream from 0 */
+	if (ret != 0) {
+		return ret;
+	}
+
+	/* gives the request ID */
+	return dma_start(data->dma_rx.dma_dev, data->dma_rx.channel);
+}
+
+static int spi_mcux_dma_move_buffers(const struct device *dev, size_t len,
+			const struct spi_config *spi_cfg, bool last_packet)
+{
+	struct spi_mcux_data *data = dev->data;
+	int ret;
+
+	ret = spi_mcux_dma_rx_load(dev, data->ctx.rx_buf, len);
+
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = spi_mcux_dma_tx_load(dev, data->ctx.tx_buf, spi_cfg,
+							len, last_packet);
+
+	return ret;
+}
+
+static int wait_dma_rx_tx_done(const struct device *dev)
+{
+	struct spi_mcux_data *data = dev->data;
+	int ret = -1;
+
+	while (1) {
+		ret = spi_context_wait_for_completion(&data->ctx);
+		if (data->status_flags & SPI_MCUX_FLEXCOMM_DMA_ERROR_FLAG) {
+			return -EIO;
+		}
+
+		if ((data->status_flags & SPI_MCUX_FLEXCOMM_DMA_DONE_FLAG) ==
+			SPI_MCUX_FLEXCOMM_DMA_DONE_FLAG) {
+			return 0;
+		}
+	}
+}
+
+static int transceive_dma(const struct device *dev,
+		      const struct spi_config *spi_cfg,
+		      const struct spi_buf_set *tx_bufs,
+		      const struct spi_buf_set *rx_bufs,
+		      bool asynchronous,
+		      struct k_poll_signal *signal)
+{
+	const struct spi_mcux_config *config = dev->config;
+	struct spi_mcux_data *data = dev->data;
+	SPI_Type *base = config->base;
+	int ret;
+	uint32_t word_size;
+
+	spi_context_lock(&data->ctx, asynchronous, signal);
+
+	ret = spi_mcux_configure(dev, spi_cfg);
+	if (ret) {
+		goto out;
+	}
+
+	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, 1);
+
+	spi_context_cs_control(&data->ctx, true);
+
+	word_size = SPI_WORD_SIZE_GET(spi_cfg->operation);
+
+	data->dma_rx.dma_cfg.dest_data_size = (word_size > 8) ?
+				(sizeof(uint16_t)) : (sizeof(uint8_t));
+	data->dma_tx.dma_cfg.dest_data_size = data->dma_rx.dma_cfg.dest_data_size;
+
+	while (data->ctx.rx_len > 0 || data->ctx.tx_len > 0) {
+		size_t dma_len;
+		bool last = false;
+
+		if (data->ctx.rx_len == 0) {
+			dma_len = data->ctx.tx_len;
+			last = true;
+		} else if (data->ctx.tx_len == 0) {
+			dma_len = data->ctx.rx_len;
+			last = true;
+		} else if (data->ctx.tx_len == data->ctx.rx_len) {
+			dma_len = data->ctx.rx_len;
+			last = true;
+		} else {
+			dma_len = MIN(data->ctx.tx_len, data->ctx.rx_len);
+			last = false;
+		}
+
+		data->status_flags = 0;
+
+		ret = spi_mcux_dma_move_buffers(dev, dma_len, spi_cfg, last);
+		if (ret != 0) {
+			break;
+		}
+
+		ret = wait_dma_rx_tx_done(dev);
```

Comment:
this will keep the CPU in while loop without task switch, it is not the purpose of DMA transfer. suggest to use the spi_context_wait_for_completion to wait for a signal.

Reply:
`wait_dma_rx_tx_done` calls `spi_context_wait_for_completion`. 
https://github.com/zephyrproject-rtos/zephyr/pull/28973/files/9e0ea17fa05989a21b5422bad2746acc5de5efa2#diff-a7245560d6f5d378f646c701cb6046fcR518

Reply:
right. I miss this.